sympy/solvers/ode/ode.py

```diff
@@ -706,6 +706,222 @@ def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):
             rv = rv1
     return rv
 
+def dsubs(eq, trans, newvars=None, reverse=False):
+    r"""
+    Function that substitutes for variables in an Ordinary Differential Equation.
+
+    The inputs are -
+
+    ``trans``: A dictionary specifying the variable transformations
+    with the old variable as key and the transformation as value.
+
+    ``newvars``: A list of new variables in the transformation rules.
+    It is required if the number of new variables is not the same as
+    the number of old variables.
+
+    ``reverse``: A boolean variable that indicates if the transformation
+    is being done in the reverse direction.
+
+    Returns
+    =======
+
+    The transformed differential equation
+
+    Examples
+    ========
+
+    A basic transformation can be achieved by -
+
+    >>> from sympy import symbols, dsubs, Function, cos
+    >>> f, g, h = symbols('f g h', cls=Function)
+    >>> x, a, t = symbols('x a t')
+    >>> eq = f(x).diff(x) + f(x)
+    >>> dsubs(eq, {x: t**2})
+    f(t**2) + Derivative(f(t**2), t)/(2*t)
+
+    Functions in the equation can be transformed too
+
+    >>> eq = f(x).diff(x, 2) + cos(x)
+    >>> dsubs(eq, {x: t**2, f(x): g(t)})
+    cos(t**2) + (Derivative(g(t), (t, 2))/(2*t)
+    - Derivative(g(t), t)/(2*t**2))/(2*t)
+    >>> dsubs(eq, {x: t, f(x): g(t)**2})
+    2*g(t)*Derivative(g(t), (t, 2)) + cos(t)
+    + 2*Derivative(g(t), t)**2
+
+    To apply a transformation with number of new variables greater
+    than number of old variables, a list of all new variables must be
+    given. For example, a transformation `x = a t` would be done as -
+
+    >>> dsubs(eq, {x: a*t, f(x): g(t)}, [t])
+    cos(a*t) + Derivative(g(t), (t, 2))/a**2
+    >>> dsubs(eq, {x: t, f(x): h(a)*g(t)**2}, [t, g(t)])
+    2*g(t)*h(a)*Derivative(g(t), (t, 2))
+    + 2*h(a)*Derivative(g(t), t)**2 + cos(t)
+
+    Let us see how ``dsubs`` works for some typical cases using an example
+    of homogeneous ODEs which can be solved using the substitution
+    `v = \frac{y}{x}`.
+
+    >>> from sympy import symbols, Function, dsolve
+    >>> x, t = symbols('x t')
+    >>> y, v = symbols('y v', cls=Function)
+    >>> eq = y(x).diff(x) - (x**2 + y(x)**2)/(x*y(x))
+
+    Transform the homogeneous equation to a linear equation
+
+    >>> homeq = dsubs(eq, {x:t, y(x): t*v(t)}).simplify()
+    >>> homeq
+    t*Derivative(v(t), t) - 1/v(t)
+
+    Find the solutions for the transformed equation
+
+    >>> solt = dsolve(homeq)
+    >>> solt
+    [Eq(v(t), -sqrt(C1 + 2*log(t))), Eq(v(t), sqrt(C1 + 2*log(t)))]
+
+    Re-substitute `v = y x` to get the solutions to the
+    original equation
+
+    >>> sol = list(map(lambda sol: dsubs(sol, {v(t): y(x)/x, t: x}, reverse=True), solt))
```

Comment:
Why do we need to give `reverse=True` if the dict has been reversed already?

Reply:
The `reverse=True` is required so that it just substitutes the transformation instead of trying to apply the substitution in the forward direction. Maybe we could omit this and set it to true internally if there are no `Derivative` terms in the equation (solution) we get?

Reply:
Is there any need to distinguish those cases?

I would have thought that if there are derivatives we need to transform them but there are no derivatives we don't.

Reply:
Yes, I've removed the reverse kwarg now.